### PR TITLE
Tcp Portals

### PR DESCRIPTION
**Pull request by SanjoDeundiak**

### CODE REVIEW

**Comment by SanjoDeundiak** [comment]
Merge after #1758

**Review by jared-s — CHANGES_REQUESTED** [review]
Very cool! A few questions, and a concern about Workflow changes

File: .github/workflows/rust_build_ockam_ockam_examples.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name: rust_build_ockam_ockam_examples
 on:
   workflow_dispatch:
-  push:
+  pull_request:
```
[the remark this answers is not in the document]
Comment (SanjoDeundiak) [reply .github/workflows/rust_build_ockam_ockam_examples.yml:4]: @jared-s my expectation was after merging #1758 this is gone during rebase

File: implementations/rust/ockam/ockam_examples/example_projects/ports/src/bin/inlet_hub.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        let return_route = msg.return_route();
+        let outlet_address = msg.body();
+
+        let mut inlet_route = return_route.clone();
+        inlet_route.modify().pop_back().append(outlet_address);
+
+        // let port = {
+        //     let mut rng = thread_rng();
+        //     rng.gen_range(3000..3005)
+        // };
```
[the remark this answers is not in the document]
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_examples/example_projects/ports/src/bin/inlet_hub.rs:30]: This is under active development, already have draft PR on top of it.

File: implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_sender.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    // TcpSendWorker will receive messages from the TcpRouter to send
+    // across the TcpStream to our friend
+    async fn handle_message(&mut self, ctx: &mut Context, msg: Routed<Any>) -> Result<()> {
+        // Remove our own address from the route so the other end
+        // knows what to do with the incoming message
+        let mut onward_route = msg.onward_route();
+        let recipient = onward_route.step()?;
+
+        if onward_route.next().is_ok() {
+            panic!("Should be last");
```
[the remark this answers is not in the document]
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_sender.rs]: Done

File: implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_receiver.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    // Context not to spawn a zombie task.
+    //
+    // Also: we must stop the TcpReceive loop when the worker gets
+    // killed by the user or node.
+    async fn initialize(&mut self, ctx: &mut Context) -> Result<()> {
+        let self_addr = ctx.address();
+
+        // Run in a loop until TcpWorkerPair::stop() is called
+        loop {
+            let mut buf = [0u8; 256];
```
[the remark this answers is not in the document]
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_receiver.rs]: Done

File: .github/workflows/rust_build_ockam_ockam_examples.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jared-s) [reply .github/workflows/rust_build_ockam_ockam_examples.yml:4]: Ah, thanks, I'll watch for it.

**Review by jared-s — APPROVED** [review]

**Comment by jared-s** [comment]
Merged https://github.com/ockam-network/ockam/pull/1770

**Pull request closed without merging**